Enabling Secure Boot on ODROID M1 (#593)
3mdeb/news-and-ideas: blog/content/post/2024-04-12-odroid-m1-secure-boot.md

#### Generating RSA Keys

TODO

#### Final directory structure

```shell
tree -FL 1
./
├── Linux_Upgrade_Tool/
├── rkbin/
├── u-boot/
├── keys/
└── upgrade_tool -> Linux_Upgrade_Tool/upgrade_tool
```

### Dependencies

Below are packages needed to build U-Boot on Debian 11 OS.

```shell
apt install gcc make bison flex python3 python3-dev python3-setuptools python3-pyelftools swig libssl-dev device-tree-compiler python2 bc
```

To build Rockchip U-Boot without errors we will also need
[Linaro gcc aarch64 toolchain](https://releases.linaro.org/components/toolchain/binaries/6.3-2017.05/aarch64-linux-gnu/).

### Configuration

#### rkbin

In `rkbin/RKBOOT/RK3568MINIALL.ini` update FlashBoot so it points to
`u-boot-spl.bin` file

```diff
diff --git a/RKBOOT/RK3568MINIALL.ini b/RKBOOT/RK3568MINIALL.ini
index ef6d071bc5ac..c80d817f4b94 100644
--- a/RKBOOT/RK3568MINIALL.ini
+++ b/RKBOOT/RK3568MINIALL.ini
@@ -15,7 +15,7 @@ NUM=2
 LOADER1=FlashData
 LOADER2=FlashBoot
 FlashData=bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
-FlashBoot=bin/rk35/rk356x_spl_v1.13.bin
+FlashBoot=../u-boot/spl/u-boot-spl.bin
 [OUTPUT]
 PATH=rk356x_spl_loader_v1.21.113.bin
 [SYSTEM]
```

#### U-Boot

> Update `CROSS_COMPILE_ARM64` variable inside `make.sh` with correct path to

Reviewer: @m-iwanicki please use I/We instead of you. We should write those posts in first-person form

Author: @TomaszAIR New commit is here:  af72329. I hope I managed to use I/We everywhere.

### Preparation

#### Repositories

To complete this stage I needed a couple of repositories:

* [Rockchip U-Boot](https://github.com/rockchip-linux/u-boot/tree/63c55618fbdc36333db4cf12f7d6a28f0a178017) -
This U-Boot version contains a function that saves the hash of the public key to
OTP memory, to be exact it's
[rsa_burn_key_hash](https://github.com/rockchip-linux/u-boot/blob/63c55618fbdc36333db4cf12f7d6a28f0a178017/lib/rsa/rsa-verify.c#L600).
[Hardkernel U-Boot](https://github.com/hardkernel/u-boot/tree/odroidm1-v2017.09)
also contains this functionality. I didn't test this version but it should work
with minimal changes to other steps.
* [rkbin](https://github.com/rockchip-linux/rkbin/tree/a2a0b89b6c8c612dca5ed9ed8a68db8a07f68bc0) -
Contains needed files: `Rockchip TPL`, `BL31`, `boot_merger` and `rk_sign_tool`.
* [upgrade_tool](https://github.com/hardkernel/rk3568-linux-tools/tree/1a32bc776af52494144fcef6641a73850cee628a/linux/Linux_Upgrade_Tool/Linux_Upgrade_Tool) -
We need this version of the tool because `upgrade_tool` and `rkdeveloptool`
contained in rkbin repository can't handle loaders generated with new idb
header.

> I used newest commits available in those repositories.

Author: I do, check links. `newest` is explanation why those commits in particular.

#### Generating RSA Keys and certificate

To enable Secure Boot I needed to generate RSA 2048 bit key. While SoC datasheet
says that RK3568
`Supports up to 4096 bits PKA mathematical operations for RSA/ECC` I had to use
2048 bits because it's the only key length accepted by `rsa_burn_key_hash`:

```C
if (info->crypto->key_len != RSA2048_BYTES)
  return -EINVAL;
```

> To generate RSA keys and certificate I decided to use `openssl` command.

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

compiler.

### Configuration

#### rkbin

In `rkbin/RKBOOT/RK3568MINIALL.ini` I had to update FlashBoot so it points to
`u-boot-spl.bin` file. Later this `.ini` file will be used by `boot_merger` tool
to create `rk356x_spl_loader_v1.21.113.bin` file which will contain U-Boot SPL.
It'll then be written to SPI flash memory.

```diff
diff --git a/RKBOOT/RK3568MINIALL.ini b/RKBOOT/RK3568MINIALL.ini
index ef6d071bc5ac..c80d817f4b94 100644
--- a/RKBOOT/RK3568MINIALL.ini
+++ b/RKBOOT/RK3568MINIALL.ini
@@ -15,7 +15,7 @@ NUM=2
 LOADER1=FlashData
 LOADER2=FlashBoot
 FlashData=bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
-FlashBoot=bin/rk35/rk356x_spl_v1.13.bin
+FlashBoot=../u-boot/spl/u-boot-spl.bin
 [OUTPUT]
 PATH=rk356x_spl_loader_v1.21.113.bin
 [SYSTEM]
```

#### U-Boot

I had to update cross-compiler path in `CROSS_COMPILE_ARM64` variable in
`make.sh` file in U-Boot repository so it pointed to where my cross-compiler was
installed.

```diff
-CROSS_COMPILE_ARM64=../prebuilts/gcc/linux-x86/aarch64/gcc-linaro-6.3.1-2017.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-
+CROSS_COMPILE_ARM64=/opt/toolchains/gcc-linaro-6.3.1-2017.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-
```

> ODROID-M1 uses RK3568B SoC so I used `rk3568_defconfig` configuration file as

Author: Check Odroid-M1 pictures or wiki or shop:
https://dn.odroid.com/RK3568/ODROID-M1/Docs/Rockchip_RK3568B2_Datasheet_V1.0.pdf

base.

```shell
make rk3568_defconfig
```

> SPL will store the hash of the key in OTP memory if the signature node

Author: Why? What other SPL there is at this point? `Mainline U-Boot with signature verification` is almost at the end of the blog post where I build upstream version.

containing the public key has the property `burn-key-hash = 0x1`. To add
a node with the public key to SPL we can use the `mkimage` tool. During my
first attempt, I used tool from the mainline U-Boot that I had on hand
because without changes to the configuration `mkimage` from the Rockchip
repository does not have the capability to add the key.
Unfortunately, the created signature node had an incorrect format.
I checked the contents of the signature node using the command:

```shell
fdtget -p spl/u-boot-spl.dtb /signature/key-dev
```

In the left column is signature node that was created by using `mkimage` from
mainline U-Boot and in the right column is correct signature node created
with `mkimage` built from Rockchip repository. Mainline U-Boot signature node
lacks couple of properties that `rsa_burn_key_hash` function requires
e.g. `rsa,c`.

```text
required        required
algo            algo
rsa,r-squared   rsa,np
rsa,modulus     rsa,c
rsa,exponent    rsa,r-squared
rsa,n0-inverse  rsa,modulus
rsa,num-bits    rsa,exponent-BN
key-name-hint   rsa,exponent
                rsa,n0-inverse
                rsa,num-bits
                key-name-hint
```

> To build `mkimage` that can add public key to SPL I had to set

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

signing rk356x_ramboot_loader_v1.21.108 ok
```

Loader was signed correctly even though tool printed some failures.

#### Sending ramboot loader

To send ramboot loader I again used `upgrade_tool` but this time with `db`
option.

```shell
./upgrade_tool db rkbin/rk356x_ramboot_loader_v1.21.108.bin
```

On UART console I had gotten this output

```text
Boot1 Release Time: Apr 20 2021 18:00:11, version: 1.08 USB BOOT
ChipType = 0x18, 271
SecureMode = 1
atags_set_bootdev: ret:(0)
UsbBoot ...684
powerOn 931
```

Before writing hash to OTP memory ramboot loader returned `SecureMode = 0`.

## Mainline U-Boot with signature verification

Now that we have pre-loader (TPL + SPL) signature verification we can use
pre-loader to verify next steps.

To complete this stage I decided to use v2024.01
[mainline U-Boot](https://github.com/3mdeb/u-boot/tree/2024.01-odroid-m1-sb-rk3568)
with couple changes. I also needed rkbin repository and RSA keys and certificate
(copied into u-boot directory). It's possible to use the same key as
earlier or create new one.

### U-Boot configuration

The main changes needed in my commit were just adding `signature` and
`u-boot-spl-pubkey-dtb` node in

> [`rockchip-u-boot.dtsi`](https://github.com/3mdeb/u-boot/blob/b5f0de18708112ce61a56526e6081796045e1763/arch/arm/dts/rockchip-u-boot.dtsi)

Reviewer 2: Cloud you add a DIFF here? It will be easier to read. Or, at least, don't surround the `rockchip-u-boot.dtsi` with \`\`, so, the link will be visible.

and changing `CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN` config variable to
`0x150000` to fix `alloc space exhausted` error when booting.

```text
U-Boot SPL 2024.01-dirty (Jan 01 1970 - 00:00:00 +0000)
Trying to boot from MMC2
alloc space exhausted
FIT buffer size: 1199104 bytes
Could not get FIT buffer of 1199104 bytes
        check CONFIG_SPL_SYS_MALLOC_SIZE
```

> To configure U-Boot it's enough to use `odroid-m1-sb-rk3568_defconfig` and set

Reviewer 2: Is this defconfig available publicly?

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

couple variables:

```shell
export CROSS_COMPILE=aarch64-linux-gnu-
export BL31=<path/to/rkbin>/bin/rk35/rk3568_bl31_v1.44.elf
export ROCKCHIP_TPL=<path/to/rkbin>/bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
```

This time I used `gcc-aarch64-linux-gnu` cross-compiler from Debian package
manager.

### Build

After configuration, we build by using `make`. It should build signed U-Boot with
public key embedded inside SPL.

```text
make -j$(nproc)
(...)
 Default Configuration: 'config-1'
 Configuration 0 (config-1)
  Description:  rk3568-odroid-m1.dtb
  Kernel:       unavailable
  Firmware:     atf-1
  FDT:          fdt-1
  Loadables:    u-boot
                atf-2
                atf-3
                atf-4
                atf-5
                atf-6
  Sign algo:    sha256,rsa2048:dev

>   Sign value:   c05e2589eb863bfb9b5b5c87ea9f99042c4cb78797a6cd0d772a68bfac21bc621790cb977bd414928b7b88a50273d336aea4afb7a1a4008834e1c3c18eb08fb617bb71205e6773904f42e81364a15eb5c581425f22d6be5e30dd1a44cbe7fa626b7fb34a3eb742c03b0ec47c19dd957358949764f82ffb24f9359efe083ea04ada512f8cab0f469cac28d08da3dfffbde1516cb225b3011f36495c959793d11b8c4bde234caca5ac1d779f7983dcf865be92c2ac3300e4cdf536e51fa398c8930dafbdd0a3258ba4704eebc063e5e57533a962a2da5c7eaf447b1244f9720adb3dc775d95b4c6dc99f6b9bd73ebe992095429510b86dd5ac912c24a8cec64841

Reviewer 2: Could you live a comment about this signature and signing FIT configs? The questions reader might ask are the same I asked you today.

Author: Specify which comment so I don't forget

Author: f7982c42f7d027003915f05f61168e15a1a64921

  Timestamp:    Mon May 27 15:08:11 2024
Signature written to 'u-boot.itb', node '/configurations/config-1/signature'
  OFCHK   .config
```

### Signing idbloader

Signing idbloader is similar to [Signing Loader](#signing-loader) section except
with `sb --idb` argument.
It's important to remember to sign idbloader with the same keys as used in that
section (in case current ones are different).

```shell

> tools/rk_sign_tool sb --idb ../u-boot/idbloader.img

Reviewer 2: Could this be verified without booting?

Author: f7982c42f7d027003915f05f61168e15a1a64921

********sign_tool ver 1.4********
IDB binary is ../u-boot/idbloader.img
signing idbhead...
failed to get key = sign_algo
signing idbloader ok
```

### U-Boot Verification

To check if SPL is signed correctly and that it correctly verifies U-Boot I have
written idbloader.img file to `spl` partition created in
[Writing key hash to OTP](#writing-key-hash-to-otp). I also flashed `u-boot.itb`
file to `uboot` partition.

> After writing needed files and inserting SD card into ODROID i restarted

Reviewer 2: ```suggestion
After writing needed files and inserting SD card into ODROID, I restarted
```

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

platform while keeping recovery button pressed.

```text
U-Boot SPL 2024.01-dirty (Apr 11 2024 - 09:15:28 +0200)
Trying to boot from MMC2
## Checking hash(es) for config config-1 ... sha256,rsa2048:dev+ OK
## Checking hash(es) for Image atf-1 ... sha256+ OK
## Checking hash(es) for Image u-boot ... sha256+ OK
## Checking hash(es) for Image fdt-1 ... sha256+ OK
## Checking hash(es) for Image atf-2 ... sha256+ OK
## Checking hash(es) for Image atf-3 ... sha256+ OK
## Checking hash(es) for Image atf-4 ... sha256+ OK
## Checking hash(es) for Image atf-5 ... sha256+ OK
## Checking hash(es) for Image atf-6 ... sha256+ OK
(...)
=>
```

> 

Reviewer 2: You could live here a sentence to sum up this log or to state, whether your expectations were met or not and how you made the conclusion. For example, I do not understand whether this attempt is successful or not, and why.

Author: f7982c42f7d027003915f05f61168e15a1a64921

## What's next

> While I managed enable Secure Boot on Odroid it would be good to more

Reviewer 2: ```suggestion
While I managed to enable Secure Boot on Odroid it would be good to more
```

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

> thoroughly test its security and capability.

Reviewer 2: ```suggestion
thoroughly test its security and capabilities.
```

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

Some of the questions that I would like to find answers for are whether there

> really isn't way to overwrite key hash and if it's possible to store more than

Reviewer 2: ```suggestion
really isn't any way to overwrite the hash stoored in OTP and if it's possible to store more than
```

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

one. OTP has 8k bits of memory based on RK3568 datasheet while hashes are only
256 bits in size so theoretically we could store 32 different hashes.

> Good next step would be to transfer capability of writing hash to OTP from

Reviewer 2: ```suggestion
Good next step would be to mainstream capability of writing hash to OTP from
```

Reviewer 2: Yep :sweat_smile:

Author: 051dcbef318bd22d15f61b72e7da8e782d1cd05d

Rockchip U-Boot to mainline U-Boot which would simplify whole implementation

> quite a bit.

Reviewer 2: ```suggestion
a lot.
```

---
title: Enabling Secure Boot on ODROID M1 (RK3568B)

> abstract: 'This blog post describes how to enable Secure Boot/Mode on

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

ODROID-M1(RK3568B). Read how to write hash of public key to OTP memory, how to

> sign loader and how to build signed U-Boot with enabled signature verification.'

Reviewer: @m-iwanicki you can see how @PLangowski done it in UEFI Secure Boot related post

cover: /covers/rockchip-logo.jpg
author: michal.iwanicki
layout: post
published: true
date: 2024-04-12
archives: "2024"

tags:
  - bootloader
  - secure-boot
  - rockchip
  - u-boot
categories:
  - Firmware
  - Security

---

## Introduction

Recently, I was tasked with implementing Secure Boot on the ODROID M1
(RK3568B).
In this post, I will describe how I managed to store the hash of the public key
in the OTP memory and enabled verification of the pre-loader's signature
(Rockchip TPL + U-Boot SPL) as well as the verification of the U-Boot's
configuration by SPL which includes hashes of all images contained in the FIT
image.

> ## Enabling Secure Mode

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

> Enabling Secure Mode is based on storing the hash of the public key in

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

> the OTP memory and enabling verification of the pre-loader’s signature.

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2). I mostly described how it was done on RK3568 SoC, but I added sentence with general "secure place known by Boot ROM"

After correctly executing this step, the RK3568B SoC will check if the hash

> of the public key that is contained in the booted image matches the hash

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2), rewrote it a little and added a diagram.
Even on your slide (18) there is information that it only saves hash to OTP/EFUSE.

> of the public key that is contained in the booted image matches the hash

Author: @TomaszAIR And I'm not sure if I can use/store this diagram from Rockchip. I took it from `Rockchip Secure Boot Application Note_v1.2_20160202.pdf` that's inside this [zip](https://github.com/JeffyCN/mirrors/blob/tools/linux/Linux_SecureBoot/SecureBootTool_v1.85_foruser.zip). There are also interesting diagrams in [Rockchip Secure Boot Application Note](http://resource.milesight-iot.com/files/Rockchip-Secure-Boot-Application-Note-V1.9.pdf#header-n150).
Or should I make one myself?

> of the public key that is contained in the booted image matches the hash

Reviewer 3: @m-iwanicki Since the linked document is classified as "Publicity," it indicates that it is likely meant to be shared publicly, but this doesn't always automatically grant permission for unrestricted use. What we can do is:
1. Create our own simple diagram
2. Add the annotation with the source

> of the public key that is contained in the booted image matches the hash

Reviewer 2: @m-iwanicki, have you addressed this?

> of the public key that is contained in the booted image matches the hash

Author: I guess you are right. Just checked to make sure that comments are also added in `html` page.

stored in the OTP memory and then use this key to verify whether the TPL/SPL
signatures match. If not, the image will not be launched.

> 

Author: @TomaszAIR I'm not sure what do you mean what hardware was used to verify Secure Boot? The only hardware I can think of is UART<-> USB converter to see console output. If you mean ODROID then I wrote in the beginning that I was working on Odroid-M1 (and it's in the title)

### Preparation

#### Repositories

To complete this stage I needed a couple of repositories:

* [Rockchip U-Boot](https://github.com/rockchip-linux/u-boot/) - This U-Boot
version contains a function that saves the hash of the public key to OTP memory,

> to be exact it's `rsa_burn_key_hash` in `lib/rsa/rsa-verify.c` file.

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

[Hardkernel U-Boot](https://github.com/hardkernel/u-boot/tree/odroidm1-v2017.09)
also contains this functionality. I didn't test this version but it should work
with minimal changes to other steps.
* [rkbin](https://github.com/rockchip-linux/rkbin) - Contains needed files,

> `Rockchip TPL`, `BL31`, and `boot_merger` tool used to generate loader and

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

`rk_sign_tool` used to sign loader
* [upgrade_tool](https://github.com/hardkernel/rk3568-linux-tools/tree/master/linux/Linux_Upgrade_Tool/Linux_Upgrade_Tool) -
We need this version of the tool because `upgrade_tool` and `rkdeveloptool`
contained in rkbin repository can't handle loaders generated with new idb
header.

I used the newest commits in default branches. Below are hashes of commits used.

> * U-Boot - 63c55618fbdc36333db4cf12f7d6a28f0a178017

Author: I changed links to point directly to commits but still to Rockchip repositories. If it's needed then I guess we could fork it. [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f)

* rkbin - a2a0b89b6c8c612dca5ed9ed8a68db8a07f68bc0
* upgrade_tool - 1a32bc776af52494144fcef6641a73850cee628a

> #### Generating RSA Keys and certificate

Author: I did not test others but I added information that the code that writes hash into memory expect 2048 bits.

> #### Generating RSA Keys and certificate

Reviewer 2: @m-iwanicki, can you please link the commit where the changes were added?

> #### Generating RSA Keys and certificate

Author: https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2#diff-101a8bd2e875f2de38635cd8af2a9dea3e40290b6e7bbfed20a6be1206993eb1R81

Keys can be generated with `openssl` command.

```shell
openssl genrsa -out keys/dev.key 2048
openssl rsa -in keys/dev.key -pubout -out keys/dev.pubkey
```

Or by using `rk_sign_tool` from rkbin repository.

```shell
rkbin/tools/rk_sign_tool kk --bits=2048 --out keys
mv keys/private_key.pem keys/dev.key
mv keys/public_key.pem keys/dev.pubkey
```

I used `openssl` to generate certificate

```shell
openssl req -batch -new -x509 -key keys/dev.key -out keys/dev.crt
```

> #### Final directory structure

Author: Added `Plan` section with short description of steps needed. [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f)

This is my final directory structure that contains needed repositories and
tools. I also created a symlink to upgrade_tool to make commands shorter.

```shell
tree -FL 1
./
├── Linux_Upgrade_Tool/
├── rkbin/
├── u-boot/
├── keys/
└── upgrade_tool -> Linux_Upgrade_Tool/upgrade_tool
```

#### Dependencies

Below are packages needed to build U-Boot on Debian 11 (bullseye) OS.

```shell

> apt install gcc make bison flex libncurses-dev python3 python3-dev python3-setuptools python3-pyelftools swig libssl-dev device-tree-compiler python2 bc

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2), added line break

```

To build Rockchip U-Boot I also needed cross-compiler. In `make.sh` file u-boot
uses Linaro 6.3.1 toolchain. At first, I tried to use cross-compiler installed
from apt package manager but unfortunately build ended in errors. Fixing one
error led to another so I chose to use
[Linaro](https://releases.linaro.org/components/toolchain/binaries/6.3-2017.05/aarch64-linux-gnu/)
compiler.

### Configuration

#### rkbin

> In `rkbin/RKBOOT/RK3568MINIALL.ini`   I had to update FlashBoot so it points to

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2)

`u-boot-spl.bin` file

> ```diff

Author: Changed in [9605d08](https://github.com/3mdeb/news-and-ideas/pull/593/commits/9605d0863b5fa9a1961c20d7812398b0388ca4f2). Added short explanation.

diff --git a/RKBOOT/RK3568MINIALL.ini b/RKBOOT/RK3568MINIALL.ini
index ef6d071bc5ac..c80d817f4b94 100644
--- a/RKBOOT/RK3568MINIALL.ini
+++ b/RKBOOT/RK3568MINIALL.ini
@@ -15,7 +15,7 @@ NUM=2
 LOADER1=FlashData
 LOADER2=FlashBoot
 FlashData=bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
-FlashBoot=bin/rk35/rk356x_spl_v1.13.bin
+FlashBoot=../u-boot/spl/u-boot-spl.bin
 [OUTPUT]
 PATH=rk356x_spl_loader_v1.21.113.bin
 [SYSTEM]
```

#### U-Boot

I had to update cross-compiler path in `CROSS_COMPILE_ARM64` variable in
`make.sh` file in U-Boot repository so it pointed to where my cross-compiler was
installed.

```diff
-CROSS_COMPILE_ARM64=../prebuilts/gcc/linux-x86/aarch64/gcc-linaro-6.3.1-2017.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-
+CROSS_COMPILE_ARM64=/opt/toolchains/gcc-linaro-6.3.1-2017.05-x86_64_aarch64-linux-gnu/bin/aarch64-linux-gnu-
```

ODROID-M1 uses RK3568B SoC so I used `rk3568_defconfig` configuration file as
base.

```shell

> make rk3568_defconfig

Author: This make is to copy configuration file (it's in configuration section). Later in `Building U-Boot` section I use `make.sh` to build U-Boot

  │   Location:                                                            │
  │     -> Boot images                                                     │
  │ (2)   -> Support Flattened Image Tree (FIT [=y])                       │
  │   Defined at Kconfig:309                                               │
  │   Depends on: FIT [=y] && SPL [=y] && SPL_DM [=y]                      │
  │   Selects: SPL_FIT [=y] && SPL_RSA [=y]                                │
  ├────────────────────────────────────────────────────────────────( 99%)──┤
  │                                < Exit >                                │
  └────────────────────────────────────────────────────────────────────────┘
```

### Building U-Boot

I used `make.sh` script to build U-Boot. Before setting `FIT_SIGNATURE` and
`SPL_FIT_SIGNATURE` variables build completed successfully, after setting those
variables `make.sh` script ends in error but fortunately everything I needed
was built correctly. Build output should show
`Platform RK3568 is build OK, with exist .config` message along with error:

* When there are no keys in `u-boot/keys` directory

  ```text
  ERROR: No keys/dev.key
  ```

> * When there are keys in `u-boot/keys` directory

Author: Yes I'm sure because that's how I did this. Error only happens after needed files are built I think during signing process. It's not needed because later I do it manually.

  ```text
  Failed: external offset 0x1000 overlaps FIT length 0x1200
  ./tools/mkimage Can't add hashes to FIT blob: -22
  ```

The files we need are `spl/u-boot-spl.dtb` and `u-boot.itb`.

Now it's time to add public key to u-boot-spl.dtb

> ```shell

Author: I shortened output a little in [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f)

 Image 1 (atf-1)
  Description:  ARM Trusted Firmware
  Created:      Fri Apr 12 13:02:03 2024
  Type:         Firmware
  Compression:  gzip compressed
  Data Size:    57097 Bytes = 55.76 KiB = 0.05 MiB
  Architecture: AArch64
  Load Address: 0x00040000
  Hash algo:    sha256
  Hash value:   6b7aa1822c907c53c1dbd8f6afe1f25c62592f940fd853892ed3b1a64afabbc0
(...)
 Default Configuration: 'conf'
 Configuration 0 (conf)
  Description:  rk3568-evb
  Kernel:       unavailable
  Firmware:     atf-1
  FDT:          fdt
  Loadables:    uboot
                atf-2
                atf-3
                atf-4
                atf-5
                atf-6
                optee
```

To verify whether signature node with public key was added to SPL we can use
`fdtget` command that I used [before](#u-boot).

Now that I had SPL with public key I needed to add `burn-key-hash` property
to this node. When SPL sees this property it'll try to write hash of public key
to OTP memory.

```shell
fdtput -tx spl/u-boot-spl.dtb /signature/key-dev burn-key-hash 0x1
```

After that I created new `u-boot-spl.bin` file

```shell
cat spl/u-boot-spl-nodtb.bin spl/u-boot-spl.dtb > spl/u-boot-spl.bin
```

> ### Creating Loader

Author: Written short Plan at the beginning in [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f)

To create loader I used `boot_merger` tool from rkbin repository. Loader that
was created when building u-boot contains old SPL without signature node so I
needed to create new one. To do that I used `RKBOOT/RK3568MINIALL.ini` config
file that was modified in [configuration](#rkbin) step

```shell
tools/boot_merger RKBOOT/RK3568MINIALL.ini
********boot_merger ver 1.34********
Info:Pack loader ok.
```

There should now be `rk356x_spl_loader_v1.21.113.bin` file in rkbin folder.

### Sending loader to ODROID

To update/write SPL to SPI flash memory I needed to enter MaskROM mode on
ODROID. Easiest way to do that is to restart device while pressing recovery
button. This way ODROID will try to load SPL from eMMC/SD memory. If there is
no eMMC/SD connected then platform will enter MaskROM mode.

> #### Clearing SPI

Author: It's written in first 2 sentences that while most likely this step isn't needed I can't be sure as I had completely cleared SPI memory. I would need to flash original SPI flash contents and test if following steps can be completed successfully.

This step could most likely be skipped. I'll describe it because during my tries
to enable Secure Boot I cleared SPI memory multiple times.
I used `upgrade_tool` from hardkernel.

```shell
sudo ./upgrade_tool ef rkbin/rk356x_spl_loader_v1.21.113.bin
Using /home/user/odroid/Linux_Upgrade_Tool/config.ini
Program Log will save in the /root/upgrade_tool/log
Loading loader...
Erase flash ok.
```

On UART console there should be output from loader:

```text
Boot1 Release Time: Apr 14 2023 10:04:54, version: 1.17
support nand flash type: slc
...nandc_flash_init enter...
No.1 FLASH ID:ff ff ff ff ff ff
sfc nor id: c2 25 38
read_lines: 2
UsbBoot ...8851
powerOn 906
```

Now restart platform and enter MaskROM mode again.

#### Upgrading loader

> Upgrading loader was very easy and was done in just one command.

Author: Changed a little in [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f). We are upgrading pre-loader which is TPL + SPL.

```shell
sudo ./upgrade_tool ul rkbin/rk356x_spl_loader_v1.21.113.bin
Using /home/user/odroid/Linux_Upgrade_Tool/config.ini
Program Log will save in the /root/upgrade_tool/log
Loading loader...
Support Type:RK3568    Loader ver:1.01 Loader Time:2024-04-11 12:37:19
Upgrade loader ok.
```

### Writing key hash to OTP

If there was U-Boot in SPI memory then SPL should try to boot it and should
write key hash into OTP memory and write `RSA: Write RSA key hash successfully.`
to UART console if writing was successfull.

In my case there was nothing in SPI flash so I had to also flash U-Boot image.
I decided to do it on SD card. To do that I created 3 partitions:

|  Partlabel   | Starting sector |  size  |
|--------------|-----------------|--------|
|      spl     |         64      |   4M   |
|    uboot     |      16384      |   4M   |
|     misc     |      24576      |   4M   |

> `spl` partition will be used later to write `idbloader.img`. SPL needs `uboot`

Author: Changed description a little earlier that SPL only writes hash after encountering `burn-key-hash` property which will only happen when trying to verify U-Boot which will only happen if U-Boot exists

and `misc` partition to correctly try to boot U-Boot and write key hash to OTP
memory.

I flashed `u-boot.itb` image to `uboot` partition and left `mics` partition
empty.

```shell
sudo dd if=u-boot.itb of=/dev/disk/by-partlabel/uboot
```

After inserting SD card into ODROID-M1 and restarting it I got this output on
UART console:

```text
U-Boot SPL board init
U-Boot SPL 2017.09-g63c55618fb-240223-dirty #iwans (Apr 11 2024 - 14:53:32)
Trying to boot from MMC2
No bootable slots found, use lastboot.
Trying fit image at 0x4000 sector
## Verified-boot: 0

sha256,rsa2048:dev
## Verified-boot: 0
RSA: Write RSA key hash successfully.
+
```

From this moment ODROID stopped booting any image that wasn't signed with
correct keys.

### Verification

To verify whether platform really verifies signatures I tried to run unsigned
loader/pre-loader and also ones signed with wrong keys.

#### Generating ramboot loader

To check SecureMode state we need to run ramboot loader. To do that I used
`boot_merger` tool with `RK3568MINIALL_RAMBOOT.ini` config file

> ```shell

Author: Clarified in [89ee811](https://github.com/3mdeb/news-and-ideas/pull/593/commits/89ee811f5893765bc657e99ba5a13f39e344ecee)

tools/boot_merger RKBOOT/RK3568MINIALL_RAMBOOT.ini
```

#### Signing loader

Now we need to sign generated `rk356x_ramboot_loader_v1.21.108.bin` loader.
To do that I used `rk_sign_tool` from rkbin repository.
First I needed to configure this tool with correct SoC and keys

```shell
 tools/rk_sign_tool cc --chip 3568 && tools/rk_sign_tool lk --key ../keys/dev.key --pubkey ../keys/dev.pubkey
********sign_tool ver 1.4********
set chip is 3568
setting chip ok.
********sign_tool ver 1.4********
private key is ../keys/dev.key
public key is ../keys/dev.pubkey
loading key ok.
```

After which I signed loader.

```shell
tools/rk_sign_tool sl --loader rk356x_ramboot_loader_v1.21.108.bin
********sign_tool ver 1.4********
Loader is rk356x_ramboot_loader_v1.21.108.bin
signing usbhead...
failed to get key = sign_algo
signing flashhead...
failed to get key = sign_algo
signing rk356x_ramboot_loader_v1.21.108 ok
```

#### Sending ramboot loader

Now I sent ramboot loader to ODROID.

> ```shell

Author: [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f): Clarified a little earlier that it's TPL + `rk356x_ramboot_loader_v1.21.108.bin` file.

./upgrade_tool db rkbin/rk356x_ramboot_loader_v1.21.108.bin
```

On UART console I had gotten this output

```text
Boot1 Release Time: Apr 20 2021 18:00:11, version: 1.08 USB BOOT
ChipType = 0x18, 271
SecureMode = 1
atags_set_bootdev: ret:(0)
UsbBoot ...684
powerOn 931
```

Before writing hash to OTP memory ramboot loader returned `SecureMode = 0`.

## Mainline U-Boot with signature verification

Now that we have pre-loader (TPL + SPL) signature verification we can use
pre-loader to verify next steps.

To complete this stage I decided to use
[mainline U-Boot](https://github.com/u-boot/u-boot). I also needed rkbin
repository and RSA keys and certificate. It's possible to use the same key as
earlier or create new one.

Keys and certificates should be copied to u-boot directory.
I checked out tag `v2024.01` in U-Boot repository.

### U-Boot configuration

To build U-Boot I used `odroid-m1-rk3568_defconfig` configuration. The only

> change I needed to make to this configuration was changing

Author: I didn't debug this too deeply so I can't say much more than what's can be glanced from error itself which is it tries to allocate more memory than it's allowed. Maybe it needs to load whole FIT image before verifying/calculating hashes/signatures and with added signature it's too big.

`CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN` to `0x150000`. Without this change
it's impossible to boot U-Boot:

```text
U-Boot SPL 2024.01-dirty (Jan 01 1970 - 00:00:00 +0000)
Trying to boot from MMC2
alloc space exhausted
FIT buffer size: 1199104 bytes
Could not get FIT buffer of 1199104 bytes
        check CONFIG_SPL_SYS_MALLOC_SIZE
```

To add signature verification I needed to add `signature` node in configuration
node and also `u-boot-spl-pubkey-dtb` node to automatically add public key
to SPL. Changes needed to `arch/arm/dts/rockchip-u-boot.dtsi` file are below

> ```diff

Author: Fixed in [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f).

  Timestamp:    Tue Apr 16 17:43:41 2024
Signature written to 'u-boot.itb', node '/configurations/config-1/signature'
```

### Signing idbloader

To sign idbloader I used once again `rk_sign_tool` from rkbin repository.
Keys used to sign idbloader need to be the same that we used with Rockchip
U-Boot in previous stage.

```shell
tools/rk_sign_tool sb --idb ../u-boot/idbloader.img
********sign_tool ver 1.4********
IDB binary is ../u-boot/idbloader.img
signing idbhead...
failed to get key = sign_algo
signing idbloader ok
```

### U-Boot Verification

To check if SPL is signed correctly and that it correctly verifies U-Boot I have
written idbloader.img file to `spl` partition created in
[Writing key hash to OTP](#writing-key-hash-to-otp). I also flashed `u-boot.itb`
file to `uboot` partition.

After writing needed files and inserting SD card into ODROID i restarted
platform while keeping recovery button pressed.

> ```text

Reviewer: @m-iwanicki do not you boot up until U-Boot shell? We could show that

Author: In [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f) I added

```
(...)
=>
```

Whole output would be too long

U-Boot SPL 2024.01-dirty (Apr 11 2024 - 09:15:28 +0200)
Trying to boot from MMC2
## Checking hash(es) for config config-1 ... sha256,rsa2048:dev+ OK
## Checking hash(es) for Image atf-1 ... sha256+ OK
## Checking hash(es) for Image u-boot ... sha256+ OK
## Checking hash(es) for Image fdt-1 ... sha256+ OK
## Checking hash(es) for Image atf-2 ... sha256+ OK
## Checking hash(es) for Image atf-3 ... sha256+ OK
## Checking hash(es) for Image atf-4 ... sha256+ OK
## Checking hash(es) for Image atf-5 ... sha256+ OK
## Checking hash(es) for Image atf-6 ... sha256+ OK

> ```

Reviewer: @m-iwanicki summary would be nice here, also some next steps. Some ideas

* integration with Yocto
* contribution to U-Boot mainline so the Rockchip one will not be needed (maybe mainline version for some reason does not want to support OTP functionality)
* understand why there is multiple versions of rkbin tools, where some works and others don't?

Author: I added short `What's next` in [11a1594](https://github.com/3mdeb/news-and-ideas/pull/593/commits/11a15940ed80c91310292ac7654018debeb2a87f)
